README.md
# linkmon
> NPM link monitor CLI utility

![linkmon monitoring directories](screenshot.jpg)

## About

Monitor and display NPM-linked project dependencies in the terminal. Watches directories and displays linked projects so you can keep track of what is linked where.

Why? I frequently find that some operations result in npm stripping my linked modules and replacing them with copies from the registry, which can be very frustrating during development across multiple libraries. This project provides aid by showing you all links so you can keep track of what is linked where.

**`linkmon` can also notify you when a linked dependency becomes unlinked!**

## Installation

Install by running `npm install -g linkmon`, if you intend on using it as a CLI utility. Install it for use within a project by running `npm install linkmon --save`.

## Usage

`linkmon` id designed as a CLI utility first:

```
Usage: linkmon [options] [arguments]

    Watch a directory:
        linkmon ~/programming/projects
    Watch multiple directories, with notifications:
        linkmon /Users/lucy/git ~/temp/my-project --notify
    Custom watch interval (30 seconds):
        linkmon some-dir --interval=30

Options:
    --interval=             Watch interval, in seconds, for monitoring
                            the provided project directories. Defaults
                            to 15 seconds.
    --notify                Enable operating-system notifications. Is
                            disabled by default. Requires a GUI such
                            as Mac OS or Windows.

Arguments:
    Provide one or more directories for scanning and monitoring. A
    directory can either contain one or more further directories that
    house NodeJS projects, or can be a NodeJS project itself. A
    directory is classified as a NodeJS project if it contains a
    package.json file and a node_modules directory.
```

### NodeJS API

`linkmon` can be used as a NodeJS API as well:

```javascript
const { scanLinks } = require("linkmon");

(async function() {
    const links = await scanLinks("../");
    // links look like the following:
    // {
    //     parent: "my-lib",
    //     parentVersion: "0.3.4",
    //     linkedPackage: "@company/package",
    //     linkedPackageVersion: "5.3.12",
    //     link: "/Users/joe/work/package-dev",
    //     path: "/Users/joe/work/my-lib"
    // }
})();
```


CHANGELOG.md
# LinkMon Changelog

## v0.1.1
_2019-10-07_

 * **Bugfix:**
   * Relative directories would throw an error

## v0.1.0
_2019-10-07_

 * Initial release


source/scan.js
const path = require("path");
const fs = require("fs");
const fileExists = require("file-exists");
const directoryExists = require("directory-exists");
const pify = require("pify");

const readdir = pify(fs.readdir);
const realpath = pify(fs.realpath);

async function getProjectLinks(projectDirectory) {
    const nodeModules = path.join(projectDirectory, "./node_modules");
    const pkgJsonPath = path.join(projectDirectory, "./package.json");
    const pkgJson = require(pkgJsonPath);
    const links = [];
    const dirContents = await readdir(nodeModules, { withFileTypes: true });
    const additionalScopedContents = [];
    for (const fileItem of dirContents) {
        if (/^@[a-zA-Z]/.test(fileItem.name)) {
            const additionalItems = await readdir(path.join(nodeModules, fileItem.name), { withFileTypes: true });
            additionalScopedContents.push(
                ...additionalItems.map(item => Object.assign(item, {
                    subName: path.join(fileItem.name, item.name)
                }))
            );
        }
    }
    for (const fileItem of [...dirContents, ...additionalScopedContents]) {
        const pathName = path.join(nodeModules, fileItem.subName || fileItem.name);
        if (fileItem.isSymbolicLink()) {
            const subIsProject = await isProjectRoot(pathName);
            if (subIsProject) {
                const resolvedPath = await realpath(pathName);
                const subPkgJsonPath = path.join(pathName, "./package.json");
                const subPkgJson = require(subPkgJsonPath);
                links.push({
                    link: resolvedPath,
                    linkedPackage: subPkgJson.name,
                    linkedPackageVersion: subPkgJson.version,
                    path: pathName
                });
            } else {
                links.push({
                    linkedPackage: null,
                    path: pathName
                });
            }
        }
    }
    return links.map(link => Object.assign(link, {
        parent: pkgJson.name,
        parentVersion: pkgJson.version
    }));
}

async function isProjectRoot(directory) {
    const pkjJson = await fileExists(path.join(directory, "./package.json"));
    const nodeModules = await directoryExists(path.join(directory, "./node_modules"));
    return pkjJson && nodeModules;
}

async function scanLinks(rootDirectory) {
    const targetDirectory = path.resolve(process.cwd(), rootDirectory);
    const isProject = await isProjectRoot(targetDirectory);
    let links = [];
    if (isProject) {
        links = await getProjectLinks(targetDirectory);
    } else {
        // Perhaps a directory containing projects
        const dirContents = await readdir(targetDirectory, { withFileTypes: true });
        for (const fileItem of dirContents) {
            if (!fileItem.isDirectory()) {
                continue;
            }
            const pathName = path.join(targetDirectory, fileItem.name);
            if (await isProjectRoot(pathName)) {
                links.push(...await getProjectLinks(pathName));
            }
        }
    }
    return links;
}

module.exports = {
    scanLinks
};


source/cli.js
#!/usr/bin/env node

const chalk = require("chalk");
const columnify = require("columnify");
const minimist = require("minimist");
const keypress = require("keypress");
const notifier = require("node-notifier");
const { scanLinks } = require("./scan.js");
const { drawMain } = require("./windows.js");

const argv = minimist(process.argv.slice(2));
const {
    _: directories = [],
    help = false,
    interval: intervalRaw = 15,
    notify = false
} = argv;

// Help
if (help) {
    console.log(`
Usage: linkmon [options] [arguments]

    Watch a directory:
        linkmon ~/programming/projects
    Watch multiple directories, with notifications:
        linkmon /Users/lucy/git ~/temp/my-project --notify
    Custom watch interval (30 seconds):
        linkmon some-dir --interval=30

Options:
    --interval=             Watch interval, in seconds, for monitoring
                            the provided project directories. Defaults
                            to 15 seconds.
    --notify                Enable operating-system notifications. Is
                            disabled by default. Requires a GUI such
                            as Mac OS or Windows.

Arguments:
    Provide one or more directories for scanning and monitoring. A
    directory can either contain one or more further directories that
    house NodeJS projects, or can be a NodeJS project itself. A
    directory is classified as a NodeJS project if it contains a
    package.json file and a node_modules directory.

For further information read the documentation on GitHub:
    https://github.com/perry-mitchell/linkmon
    `.trim());
    process.exit(0);
}

// Init
if (directories.length <= 0) {
    console.error("A scan directory must be specified");
    process.exit(2);
}
const refreshInterval = parseInt(intervalRaw, 10) * 1000;
keypress(process.stdin);
process.stdin.on("keypress", function (ch, key) {
    if (key && key.name == "q") {
        process.exit(0);
    }
});
process.stdin.setRawMode(true);
process.stdin.resume();

// Execution
let previousLinks = [];
const updateContent = drawMain();
const runUpdate = () => {
    Promise
        .all(directories.map(dir => scanLinks(dir)))
        .then(linkAgg => linkAgg.reduce((output, links) => [
            ...output,
            ...links
        ], []))
        .then(links => {
            const iterableLinks = [];
            const projectLinks = links.reduce((output, link) => {
                const { parent, parentVersion } = link;
                output[parent] = output[parent] || {
                    parent,
                    parentVersion,
                    links: []
                };
                output[parent].links.push(link);
                iterableLinks.push(link);
                return output;
            }, {});
            if (notify) {
                previousLinks.forEach(previousLink => {
                    const matching = iterableLinks.find(link => link.parent === previousLink.parent && link.linkedPackage === previousLink.linkedPackage);
                    if (!matching) {
                        notifier.notify({
                            title: `Linked module removed @ ${previousLink.parent}`,
                            message: `${previousLink.linkedPackage} (${previousLink.linkedPackageVersion}) was unlinked in ${previousLink.parent} dependencies.`
                        });
                    }
                });
            }
            previousLinks = [...iterableLinks];
            let content = "";
            Object.keys(projectLinks).forEach(projectName => {
                const item = projectLinks[projectName];
                content += `${chalk.green.bold(projectName)} ${chalk.gray(`v${item.parentVersion}`)}\n`;
                content += columnify(
                    item.links.map(linkItem => ({
                        "startPadding": " ",
                        "Project": chalk.red(linkItem.linkedPackage),
                        "Version": chalk.gray(linkItem.linkedPackageVersion),
                        "arrow": chalk.dim("⟶"),
                        "Target": chalk.yellow(linkItem.link)
                    })),
                    {
                        showHeaders: false,
                        minWidth: 4
                    }
                ) + "\n";
            });
            updateContent(content);
        })
        .catch(err => {
            updateContent(`${chalk.bold.red("Error:")} ${err.message}`);
        });
};

runUpdate();
const refreshTimer = setInterval(runUpdate, refreshInterval);
